### PR TITLE
cache / sparse: next lot

### PR DESCRIPTION
**Pull request by lyakh**
next 7 commits from the cache / coherency / sparse work

### CODE REVIEW

File: zephyr/CMakeLists.txt [hunk 1]
```diff
@@ -712,6 +712,31 @@ zephyr_library_sources_ifdef(CONFIG_COMP_COPIER
...
+)
+
+zephyr_library_sources_ifdef(CONFIG_COMP_DRC
+	${SOF_AUDIO_PATH}/drc/drc.c
+	${SOF_AUDIO_PATH}/drc/drc_generic.c
+)
+
+zephyr_library_sources_ifdef(CONFIG_COMP_MULTIBAND_DRC
+	${SOF_AUDIO_PATH}/multiband_drc/multiband_drc.c
+	${SOF_AUDIO_PATH}/multiband_drc/multiband_drc_generic.c
```
Comment (marc-hb) [inline zephyr/CMakeLists.txt:737]: These really do not look like sparse fixes.

Comment (lyakh) [reply zephyr/CMakeLists.txt:737]: @marc-hb I have to update drc and other components for cache / sparse and that only works with Zephyr, so I need to add them to Zephyr builds

Comment (marc-hb) [reply zephyr/CMakeLists.txt:737]: I'm sure you know what you're doing but this seems really buried in a large PR that seems to be mostly about something else. It's not mentioned in the description and I could not find a commit message mentioning it either?

File: src/audio/smart_amp/smart_amp_generic.c [hunk 2]
```diff
@@ -80,9 +80,9 @@ static void smart_amp_s24_ff_default(const struct comp_dev *dev,
 
 #if CONFIG_FORMAT_S32LE
 static void smart_amp_s32_ff_default(const struct comp_dev *dev,
-				     const struct audio_stream *source,
-				     const struct audio_stream *sink,
-				     const struct audio_stream *feedback,
+				     const struct audio_stream __sparse_cache *source,
+				     const struct audio_stream __sparse_cache *sink,
+				     const struct audio_stream __sparse_cache *feedback,
 				     uint32_t frames)
```
Comment (dbaluta) [inline src/audio/smart_amp/smart_amp_generic.c:86]: Looks like this doesn't belong to this commit. Same with some other changes related to ```__sparse_cache```

File: zephyr/CMakeLists.txt [hunk 3]
```diff
@@ -712,6 +712,12 @@ zephyr_library_sources_ifdef(CONFIG_COMP_COPIER
 	${SOF_AUDIO_PATH}/copier.c
 )
 
+zephyr_library_sources_ifdef(CONFIG_MAXIM_DSM
+	${SOF_AUDIO_PATH}/smart_amp/smart_amp.c
+	${SOF_AUDIO_PATH}/smart_amp/smart_amp_generic.c
+	${SOF_AUDIO_PATH}/smart_amp/smart_amp_maxim_dsm.c
+)
+
```
Comment (dbaluta) [inline zephyr/CMakeLists.txt:720]: pretty sure this doesn't belong to this commit.

File: zephyr/wrapper.c [hunk 4]
```diff
@@ -600,6 +601,10 @@ int task_main_start(struct sof *sof)
 		sys_comp_crossover_init();
 	}
 
+	if (IS_ENABLED(CONFIG_COMP_DRC)) {
+		sys_comp_drc_init();
+	}
+
```
Comment (dbaluta) [inline zephyr/wrapper.c:608]: I don't understand this change. Do you enable DRC for Zephyr? Then it should be a separate commit with proper explanation.

File: src/audio/smart_amp/smart_amp_generic.c (hunk 2 above)
Comment (lyakh) [reply src/audio/smart_amp/smart_amp_generic.c:86]: @dbaluta this very much does belong to it. Maybe the commit description isn't very clear, but that's exactly what you need to do to acquire and release buffers properly: with acquired buffers you need to use the proper sparse annotation, specifically this one.

File: zephyr/CMakeLists.txt (hunk 3 above)
Comment (lyakh) [reply zephyr/CMakeLists.txt:720]: @dbaluta In principle we could separate changes to wrapper.c and to CMakeLists.c for each component, but this is a very small and simple and standard piece of code. It's very similar to what is done for all other components. No magic there at all.

Comment (dbaluta) [reply zephyr/CMakeLists.txt:720]: Sure, maybe add some comments inside the commit message. Saying that we also enable the component for zephyr?

Comment (lgirdwood) [reply zephyr/CMakeLists.txt:720]: lets make this another commit.

File: zephyr/CMakeLists.txt [hunk 5]
```diff
@@ -722,6 +722,11 @@ zephyr_library_sources_ifdef(CONFIG_COMP_ARIA
 	${SOF_AUDIO_PATH}/aria/aria.c
 )
 
+zephyr_library_sources_ifdef(CONFIG_COMP_CROSSOVER
+	${SOF_AUDIO_PATH}/crossover/crossover.c
+	${SOF_AUDIO_PATH}/crossover/crossover_generic.c
+)
+
```
Comment (lgirdwood) [inline zephyr/CMakeLists.txt:729]: New commit,.

**Review by lgirdwood — COMMENTED** [review]
Need to make the Zephyr build updates as another patch or PR. Sparse changes look good.

**Comment by lyakh** [comment]
CI: only one suspend-resume failure https://sof-ci.01.org/sofpr/PR5957/build787/devicetest/?model=WHL_UPEXT_HDA_ZEPHYR&testcase=check-suspend-resume-with-capture-5

**Comment by lgirdwood** [comment]
@lyakh two things
1) Can you prep the next sparse PR for review (we need to get to the point where we can start CI using Sparse soon).
2) Can you create an issue for the Zephyr linker - your workaround is fine, but long term we need t use the ELF section like native SOF for all the processing modules. Thanks.
3) We need to start using sparse for multicore cache/uncache objects in more places (including Zephyr)

**Comment by lyakh** [comment]
> * Can you prep the next sparse PR for review (we need to get to the point where we can start CI using Sparse soon).

@lgirdwood #5976 

> * Can you create an issue for the Zephyr linker - your workaround is fine, but long term we need t use the ELF section like native SOF for all the processing modules. Thanks.

For the record: it isn't my workaround. It's been there from day one of SOF Zephyr port. Yes, we have to do that, it should be somewhere in our plans...

> * We need to start using sparse for multicore cache/uncache objects in more places (including Zephyr)

Yes, but maybe (1) merge the rest of `comp_buffer` sharing in #5976, (2) collect some experience with that, (3) benchmark performance differences with `CONFIG_INCOHERENT` disabled.
